Change application URLs to allow remote browsing

zack4485 opened the pull request:
If I do `dotnet run` or `dotnet build` on my machine and I then want to browse the page from a different device (ex: my iPhone connected to the same wifi network as my laptop) then I need the application URLs to use 0.0.0.0 (essentially wildcard) instead of localhost.

zack4485 committed:
Change application URLs to allow remote browsing

If I do `dotnet run` or `dotnet build` on my machine and I then want to browse the page from a different device (ex: my iPhone connected to the same wifi network as my laptop) then I need the application URLs to use 0.0.0.0 (essentially wildcard) instead of localhost.

mythz commented:
I've reverted https back to localhost since `0.0.0.0` was failing to load in all my browsers due to security warnings.

There's no security issues running on `http:0.0.0.0` so I've changed launchSettings to run on both `https://localhost:5002;http://0.0.0.0:5000` in this commit:

https://github.com/NetCoreTemplates/razor-ssg/commit/400a60204fd0a822fba560603b9b5466626b7d7d

zack4485 commented:
I’m not at my computer right now but I’m pretty sure most browsers will
attempt to automatically redirect a user from http to https even if the
server itself doesn’t do it.  From memory, when you configure launch
settings with https://localhost this outright prevents the web server from
listening for https requests on the machine’s up address or the loopback
address.

In other words I think the combination of browsers trying to protect users
(by attempting https requests even when the user specifies http in the url)
and not having https://0.0.0.0 in the launch settings file will mean users
attempting to remotely browse the kestrel/iis server during debug sessions
will get an http error.

Methinks the correct fix to satisfy all parties—namely your desire to avoid
getting security warnings—would be to specify the following:

https://localhost:5002;http://0.0.0.0:5000;https://0.0.0.0:5002

What say you?

On Wed, Oct 30, 2024 at 9:51 PM Demis Bellot ***@***.***>
wrote:

> I've reverted https back to localhost since 0.0.0.0 was failing to load
> in all my browsers due to security warnings.
>
> There's no security issues running on http:0.0.0.0 so I've changed
> launchSettings to run on both https://localhost:5002;http://0.0.0.0:5000
> in this commit:
>
> 400a602
> <https://github.com/NetCoreTemplates/razor-ssg/commit/400a60204fd0a822fba560603b9b5466626b7d7d>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NetCoreTemplates/razor-ssg/pull/5#issuecomment-2448855324>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARIRKA6KMYL2KWOX2FR3JWDZ6GEKJAVCNFSM6AAAAABQ46L3ZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINBYHA2TKMZSGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mythz commented:
You can't have multiple bindings on the same port and server redirection is not enabled during development. Browsers shouldn't disobey an explicit `http://0.0.0.0` request and even if they attempted SSL first they would fallback to HTTP. They might flag http:// requests as non secure but I've never had issues trying to browse any local http servers.

Don't want to add another binding since these 2 should justify most use cases for development. Devs who need it can modify their local copy to change their bindings.

Also please ensure every PR results doesn't break the build, i.e. it's better to have fewer PRs that are locally tested and built before they're submitted.

zack4485 commented:
Yep I know multiple bindings can’t share the same port; that was a typo in my email.

The specific problem with what you’re proposing is that the use of non-standard ports (5000/5002) mean the browser-level fallback to HTTP doesn’t work; if a user types in https://10.0.0.2:5002 it’ll fail to load (because Kestrel only allows localhost) and the browser doesn’t automatically fall back to port 5000 for non-SSL connections.

For what it’s worth I think this could be solved for by building a custom wildcard cert for the Kestrel server and specifying it in appsettings.  https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints?view=aspnetcore-8.0#configure-https-in-appsettingsjson

This isn’t a big enough deal to do battle over, but just wanted to explain an alternative and why I’m asking for it.

Sorry I broke the build!  It did build locally before I did the commits but I missed one of the changes and I got lazy at 3AM and didn’t test the aggregate behavior of the PRs.  Won’t happen again.

Thanks
Zach


From: Demis Bellot ***@***.***>
Date: Thursday, October 31, 2024 at 12:03 PM
To: NetCoreTemplates/razor-ssg ***@***.***>
Cc: zack4485-MSFT ***@***.***>, Author ***@***.***>
Subject: Re: [NetCoreTemplates/razor-ssg] Change application URLs to allow remote browsing (PR #5)

You can't have multiple bindings on the same port and server redirection is not enabled during development. Browsers shouldn't disobey an explicit http://0.0.0.0 request and even if they attempted SSL first they would fallback to HTTP. They might flag http:// requests as non secure but I've never had issues trying to browse any local http servers.

Don't want to add another binding since these 2 should justify most use cases for development. Devs who need it can modify their local copy to change their bindings.

Also please ensure every PR results doesn't break the build, i.e. it's better to have fewer PRs that are locally tested and built before they're submitted.

—
Reply to this email directly, view it on GitHub<https://github.com/NetCoreTemplates/razor-ssg/pull/5#issuecomment-2450253510>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARIRKA4LTZYZEUMSXML7ZADZ6JIG3AVCNFSM6AAAAABQ46L3ZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJQGI2TGNJRGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

mythz commented:
Yep understood, whilst you wouldn't be able to use `https://10.0.0.2:5002` you should be able to use `http://10.0.0.2:5000` without issue. 

The template should just aim to have a broadly useful default configuration and a `https://0.0.0.0:5002` wildcard binding broke the most popular dev use case of viewing it in your local browser. Also don't want it to listen on more than 2 bindings by default and only want to depend on the default dev cert in the default template. Anyone needing to, can modify they're localSettings to add/modify any custom certs or additional bindings as they wish.
